python/pyarrow/lib.pxd

```diff
@@ -142,6 +142,21 @@ cdef class PyExtensionType(ExtensionType):
     pass
 
 
+cdef class Metadata:
+    pass
+
+
+cdef class KeyValueMetadata(Metadata):
+    cdef:
+        shared_ptr[CKeyValueMetadata] wrapped
```

Comment:
Many types return `const CKeyValueMetadata` though

Reply:
I was unsure about whether we should prefer a mutable interface and copy the underlying kvm objects or keep everything immutable. Converting it to immutable.

Reply:
Converted the python side KeyValueMetadata into an immutable object. 
Shouldn't we convert the C++ side KeyValueMetadata into an immutable construct as well?

Reply:
We handle immutability using const in C++

Reply:
@wesm the codebase rarely uses `shared_ptr<const T>`, but still most of the types provide immutable/copy-on-write interfaces.

Reply:
I meant that we use const to control for mutability specifically with KeyValueMetadata